### ISSUE TITLE
How to handle action mask in the custom network?

### ISSUE THREAD

**Issue by bbliao** [tensorforce/tensorforce#586]
Hi, thanks for the great work!
I am using the ppo agent to find a "good" path in the network: the learned agent can choose a "good" neighbor nodes given the current path (len=k) and become a new path (len=k+1) until the max_len. 
Since the number of neighbors of each node in the network is not the same, so I pad the neighbor and use action mask. Specifically, the states and actions are defined as:
```
    @property
    def states(self):
        states = dict()
        states['path_embedding'] = dict(shape=(self.dim,), type='float')
        states['neighbour_embedding'] = dict(shape=(NEIGHBOUR_NUM, self.dim), type='float')
        states['action_mask'] = dict(shape=(NEIGHBOUR_NUM,), type='float')
        return states

    @property
    def actions(self):
        return dict(type='int', num_actions=NEIGHBOUR_NUM)
``` 
where `NEIGHBOUR_NUM` denotes the maximum number (here, 27) of neighbor nodes and `action_mask` is `1` if this neighbor is available and `0` for padding.

Since the `states` contains multiple states, the custom network is used,
```
class CustomNetwork(Network):

    def tf_apply(self, x, internals, update, return_internals=False):
        path_embedding = x['path_embedding']  # 256-dim, float
        neighbour_embedding = x['neighbour_embedding']  # 27*256-dim, float
        action_mask = x['action_mask']  # 27-dim, float
        dim = 64

        # path_embedding
        path_dense_1 = tf.layers.Dense(name="path_dense_1", activation='relu', units=dim)
        path_dense_1_output = path_dense_1(path_embedding)
        path_dense_1_output_expand = tf.expand_dims(path_dense_1_output, -2)

        # neighbour_embedding
        neighbour_dense_1 = tf.layers.Dense(name="neighbour_dense_1", activation='relu', units=dim)
        neighbour_dense_1_output = neighbour_dense_1(neighbour_embedding)

        # add
        path_neighbour = tf.add(path_dense_1_output_expand, neighbour_dense_1_output)
        path_neighbour_dense_1 = tf.layers.Dense(name="path_neighbour_dense_1", activation='relu', units=dim)
        path_neighbour_dense_1_output = path_neighbour_dense_1(path_neighbour)  # 27 * 64

        path_neighbour_dense_2 = tf.layers.Dense(name="path_neighbour_dense_2", activation='relu', units=dim)
        path_neighbour_dense_2_output = path_neighbour_dense_2(path_neighbour_dense_1_output)  # 27 * 64

        path_neighbour_dense_3 = tf.layers.Dense(name="path_neighbour_dense_3", activation='relu', units=1)
        path_neighbour_dense_3_output = path_neighbour_dense_3(path_neighbour_dense_2_output)  # 27 * 1
        
        # reduce to 1-dim
        origin_output = tf.reduce_mean(path_neighbour_dense_3_output, axis=-1, name="origin_output")
        
        # add mask
        actor_output = origin_output * action_mask + (1 - action_mask) * tf.float32.min
        
        # Combination
        if return_internals:
            return actor_output, list()
        else:
            return actor_output
```
The output of this network is a vector of 27 dimension, representing the score (or "probability") of choosing the corresponding action. To avoid the agent choose the padding action (which is not valid), in the last line of the custom network, I set the probability of the padding action as 'tf.float32.min':
`actor_output = origin_output * action_mask + (1 - action_mask) * tf.float32.min`

However, the agent can still choose the padding action (which is not valid). So I dig your code and try to print what happened.
Firstly, I print the output of the custom network by adding a line in the [distribution_model.py](https://github.com/tensorforce/tensorforce/blob/520a8d992230e382f08e315ede5fc477f5e26bfb/tensorforce/models/distribution_model.py#L187):
```
def tf_actions_and_internals(self, states, internals, deterministic):
        embedding, internals = self.network.apply(
            x=states,
            internals=internals,
            update=tf.constant(value=False),
            return_internals=True
        )
        embedding = tf.Print(input_=embedding, data=[embedding], message="embedding is: ", summarize=27)
        actions = dict()  
        for name in sorted(self.distributions):
            distribution = self.distributions[name] 
            distr_params = distribution.parameterize(x=embedding)
            actions[name] = distribution.sample(
                distr_params=distr_params, 
                deterministic=tf.logical_or(x=deterministic, y=self.requires_deterministic)
            )
            actions[name] = tf.Print(input_=actions[name], data=[actions[name]], message="actions[name] is: ", summarize=27)
            ...
```
and the result makes sense, with the first k items being 0 (k available neighbors) and others being being tf.float32.min (padding action). Since the action is chosen with the embedding (output of the custom network) using the `distribution.parameterize`, so I print the intermediate results in the [categorical.py](https://github.com/tensorforce/tensorforce/blob/520a8d992230e382f08e315ede5fc477f5e26bfb/tensorforce/core/distributions/categorical.py#L53):
```
    def tf_parameterize(self, x):
        # Flat logits
        x = tf.Print(input_=x, data=[x], message="x: ", summarize=27)
        logits = self.logits.apply(x=x)
        logits = tf.Print(input_=logits, data=[logits], message="logits_before: ", summarize=27)                   

        # Reshape logits to action shape
        shape = (-1,) + self.shape + (self.num_actions,)
        logits = tf.reshape(tensor=logits, shape=shape)
        logits = tf.Print(input_=logits, data=[logits], message="logits_after: ", summarize=27)

        # !!!
        state_value = tf.reduce_logsumexp(input_tensor=logits, axis=-1)
        state_value = tf.Print(input_=state_value, data=[state_value], message="state_value: ", summarize=27)

        # Softmax for corresponding probabilities
        probabilities = tf.nn.softmax(logits=logits, axis=-1)
        probabilities = tf.Print(input_=probabilities, data=[probabilities], message="prob_before: ", summarize=27)

        # Min epsilon probability for numerical stability
        probabilities = tf.maximum(x=probabilities, y=util.epsilon)
        probabilities = tf.Print(input_=probabilities, data=[probabilities], message="prob_after: ", summarize=27)
        ...
```
the intermeidate results is:
```
embedding is: [[0 0 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38]]
x: [[0 0 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38 -3.40282347e+38]]
logits_before: [[5.53349871e+37 -1.54193044e+38 -8.09009238e+37 2.17776653e+38 -1.61143856e+38 1.28076876e+38 -2.85986602e+36 -2.41396939e+38 5.0056997e+37 2.75352918e+38 1.47537229e+38 4.49846808e+37 -3.9894663e+37 -1.0215118e+38 -1.33916982e+38 -2.06161871e+37 -3.73279926e+37 3.04329058e+38 inf 7.31203177e+37 1.13987568e+38 -1.93770049e+38 -8.56730401e+37 -1.99891077e+38 6.67217448e+37 1.14233888e+38 1.23624288e+38]]
logits_after: [[5.53349871e+37 -1.54193044e+38 -8.09009238e+37 2.17776653e+38 -1.61143856e+38 1.28076876e+38 -2.85986602e+36 -2.41396939e+38 5.0056997e+37 2.75352918e+38 1.47537229e+38 4.49846808e+37 -3.9894663e+37 -1.0215118e+38 -1.33916982e+38 -2.06161871e+37 -3.73279926e+37 3.04329058e+38 inf 7.31203177e+37 1.13987568e+38 -1.93770049e+38 -8.56730401e+37 -1.99891077e+38 6.67217448e+37 1.14233888e+38 1.23624288e+38]]
prob_before: [[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 -nan 0 0 0 0 0 0 0 0]]
prob_after: [[1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 -nan 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06 1e-06]]
actions[name] is: [4]
choose a padding station: 4/2
```
after adding the [default linear layer](https://github.com/tensorforce/tensorforce/blob/520a8d992230e382f08e315ede5fc477f5e26bfb/tensorforce/core/distributions/categorical.py#L49), the probability of the action seems confused. 

This issue is a little bit longer, thanks for your time. So, my question is, how to handle the action mask in the custom network? Or there may be some better way to handle the action mask?

**Comment by AlexKuhnle** [comment]
Hi @bbliao,

Your approach indeed doesn't work because of the linear layer in the categorical distribution module. Why? The network output is not actually the "action output", but just an arbitrarily-sized embedding (thus no need for size 27) so that users don't need to know about action-modeling, while the distribution is added afterwards to take care of action-distribution-specific things (for instance, logits for categorical actions, mean/variance for continuous actions, and more).

Action masking is actually on the roadmap for the new version, but since it may still take a while, here what you could try immediately: Implement your own `Distribution` subclass, which works like a minimal version of the existing `Categorical`, but skips the linear layer. You should then be able to specify your custom distribution via `distributions(action=dict(type=CustomCategorical))` (default action name `action`).

Let me know if it works and/or if you need more information.

(Alternatively, I could probably implement such a distribution type for the new version branch soon, but that would require you to change to the new branch.)

**Comment by bbliao** [comment]
@AlexKuhnle Thanks for your advice.
I have implemented the [CustomCategorical](https://github.com/tensorforce/tensorforce/files/3094469/custom_categorical.txt), which just comment the linear layer `self.logits` with `'''===CustomCategorical==='''` in the existing [`Categorical.py`](https://github.com/tensorforce/tensorforce/blob/master/tensorforce/core/distributions/categorical.py). 
For the use of the new `CustomCategorical` class, I add the `custom` key in the action_spec:
```
    @property
    def actions(self):
        # return dict(type='int', num_actions=NEIGHBOUR_NUM)
        return dict(type='int', num_actions=NEIGHBOUR_NUM, custom='mask')
```
and in the `create_distribution` function of the [`distribution_models.py`](https://github.com/tensorforce/tensorforce/blob/520a8d992230e382f08e315ede5fc477f5e26bfb/tensorforce/models/distribution_model.py#L160), I add the following code to use the `CustomCategorical`:
```
            elif action['type'] == 'int':
                if 'custom' in action:
                    print("distribution is CustomCategorical!")
                    distributions[name] = CustomCategorical(
                        shape=action['shape'],
                        num_actions=action['num_actions'],
                        scope=name,
                        summary_labels=self.summary_labels
                    )
                else:
                    print("distribution is Categorical!")
                    distributions[name] = Categorical(
                        shape=action['shape'],
                        num_actions=action['num_actions'],
                        scope=name,
                        summary_labels=self.summary_labels
                    )
```
However, when I run the code, I get the Nan [Error](https://github.com/tensorforce/tensorforce/files/3094540/custom_categorical.txt) (For locating the bug, I print the `logits` and `probabilities` with `tf.Print`) after 10 episodes. The update frequence of the model is also 10:
```
"update_mode": {
        "unit": "episodes",
        "batch_size": 10,
        "frequency": 10
    },
```
So I guess this Nan error may be caused by the updates of the model. Unfortunately, I have debugged this bug for hours but didn't solve it. So, could you please give me some advice? 
Thanks a lot!

**Comment by AlexKuhnle** [comment]
You don't actually need to change code in `distribution_model` and add a custom value to the actions specification, if you just use the agent argument I mentioned above -- however, I realize there is a typo in my explanation, it should be: `distributions=dict(action=dict(type=CustomCategorical))`.

But to your actual question: I expect this happens because you use negative infinity for the logits (which makes sense) but then still keep some of the calculation code of `Categorical` (which probably causes the NaNs), so here my proposal for a `CustomCategorical`, modified from yours:

```python
class CustomCategorical(Distribution):

    def __init__(self, shape, num_actions, scope='custom_categorical', summary_labels=()):
        self.num_actions = num_actions
        super(CustomCategorical, self).__init__(shape=shape, scope=scope, summary_labels=summary_labels)

    def tf_parameterize(self, x):
        logits = x
        return logits

    def state_value(self, distr_params):
        raise NotImplementedError

    def state_action_value(self, distr_params, action=None):
        raise NotImplementedError

    def tf_sample(self, distr_params, deterministic):
        logits = distr_params

        # Deterministic: maximum likelihood action
        definite = tf.argmax(input=logits, axis=-1, output_type=util.tf_dtype('int'))

        # Non-deterministic: sample action using Gumbel distribution
        uniform_distribution = tf.random_uniform(
            shape=tf.shape(input=logits),
            minval=util.epsilon,
            maxval=(1.0 - util.epsilon)
        )
        gumbel_distribution = -tf.log(x=-tf.log(x=uniform_distribution))
        sampled = tf.argmax(input=(logits + gumbel_distribution), axis=-1, output_type=util.tf_dtype('int'))

        return tf.where(condition=deterministic, x=definite, y=sampled)

    def tf_log_probability(self, distr_params, action):
        logits = distr_params
        one_hot = tf.one_hot(indices=action, depth=self.num_actions)
        return tf.reduce_sum(input_tensor=(logits * one_hot), axis=-1)

    def tf_entropy(self, distr_params):
        raise NotImplementedError

    def tf_kl_divergence(self, distr_params1, distr_params2):
        raise NotImplementedError
```

It implements only the minimally necessary without any additional computation, thus hopefully avoiding NaNs. Of course a more general version would still keep the implementations, as you did, but that requires explicitly taking care of negative infinity in `tf_parameterize`.

I hope that solves the problem!

**Comment by AlexKuhnle** [comment]
Masking now possible, will be added to documentation.
